apm-agent-core/src/main/java/co/elastic/apm/agent/bci/IndyBootstrap.java

```diff
@@ -362,10 +363,22 @@ public static ConstantCallSite bootstrap(MethodHandles.Lookup lookup,
             MethodHandle instrumentedMethod = args.length >= 5 ? (MethodHandle) args[4] : null;
 
             ClassLoader instrumentationClassLoader = ElasticApmAgent.getInstrumentationClassLoader(adviceClassName);
+            ClassLoader targetClassLoader = lookup.lookupClass().getClassLoader();
             ClassFileLocator classFileLocator;
             List<String> pluginClasses = new ArrayList<>();
             if (instrumentationClassLoader instanceof ExternalPluginClassLoader) {
-                pluginClasses.addAll(((ExternalPluginClassLoader) instrumentationClassLoader).getClassNames());
+                List<String> externalPluginClasses = ((ExternalPluginClassLoader) instrumentationClassLoader).getClassNames();
```

Comment:
Possibly outside the scope of this PR but I think we should add more validation rules in the constructor of `ExternalPluginClassLoader` to make sure external plugins don't contain slf4j and Byte Buddy, similar to how we validate that the Plugin SDK isn't packaged in the plugin.

https://github.com/elastic/apm-agent-java/blob/e756114a054faf650e92e62b1793b0a1aa996e5a/apm-agent-core/src/main/java/co/elastic/apm/agent/bci/classloading/ExternalPluginClassLoader.java#L46-L48

While looking at it, the `%s` placeholder is unresolved.

Reply:
Yes, let's handle separately (except from the Exception message String format).
With #2390 would be allowed, but all dependencies we package within the agent (not only Byte Buddy) may indeed cause `LinkageError` if used within the external plugin AND available to the CL that loads the instrumented classes.

Reply:
I was primarily thinking about `IndyPluginClassLoader` created for external plugins' instrumentation, which may cause this specific kind of `LinkageError` - "loader constraint violation" (because of the child-first nature of this CL). 
However, since the agent CL is a direct parent of the `ExternalPluginClassLoader`, packaging classes that are available within the agent may either pass quietly, or cause other types of `LinkageError` (e.g. when using mismatching versions).
So I take this back - it is within the context of this PR to disallow external plugins to package libraries that are packaged in the agent.